# PCAn_mucor 

Point Centromere Annoation for Mucoromycota (PCAn_mucor) is a tool used to identify and annotate point centromeres from the Mucoromycota clade. It is designed in Python to work on genome assemblies at any level of completeness (contig, scaffold, chromosome, or complete).

## Installation
This tool works in the following systems

<img src="https://github.com/primefaces/primeicons/blob/master/raw-svg/apple.svg" width="50" height="50"> <img src="https://upload.wikimedia.org/wikipedia/commons/thumb/3/35/Tux.svg/1280px-Tux.svg.png" width="40" height="40">

The tool was tested on the following versions
1. M1 MacBook from 2020 with MacOS Ventura 13.3.1
2. Red Hat 8.5.0-22 Linux version 4.18.0-553.22.1.el8_10.x86_64 system

Miniconda was used as the environment and package manager for this protject. To set up Miniconda, please refer to the [documentation](https://docs.anaconda.com/miniconda/index.html)

## Setting up the environment
  1. Download the [environment specifications file](https://github.com/JHelsen/point-centromere-detection/blob/main/PCAn/pcan_specs.yml)
  2. Set up and activate the environment using the following commands
     ```bash
      conda env create -n pcan --file <path/to/>pcan_specs.yml
     ```
     ```bash
      conda activate pcan
     ```

### Average installation time
The installation takes *under 5 minutes* on a system with a good (~100 Mbps) internet connection. The tool occupies 23.5 MB and the Anaconda environment occupies ~1.55 GB.

## Running PCAn_mucor
The files needed to run PCAn_mucor can be downloaded [here](https://github.com/JHelsen/point-centromere-detection/blob/main/PCAn_mucor/PCAn_mucor.zip)

After activating the pcan environment, PCAn_mucor can be run from the command line using the following command
```bash
python <path/to/>AutomatedCENretrieval_Mucor.py
```
PCAn_mucor uses [MEME](https://meme-suite.org/meme/) motifs that have been calculated for different clades within Mucoromycota to identify and extract centromeres from a given genome assembly. 

PCAn_mucor requires the following user inputs 
  1. The path to the genome assembly
  2. Selection of the motif that is phylogenetically closest to the input genome assembly

## PCAn_mucor outputs
You can find sample predictions for _Mucor circinelloides_ in the "sample_outputs" folder. The assembly for _M. circinelloides_ was obtained from [NCBI](https://www.ncbi.nlm.nih.gov/datasets/genome/GCA_037974945.1/).

## SequenceHits.fsa
This file contains the identified centromere sequences

## plots folder
The folder contains the plots generated by PCAn_mucor. Each plot is a line plot of percentage of GC content and can be used to differentiate between real and false hits. For example, in the _M. circenlloides_ analysis,

| Real hit | False hit |
| -------- | ------- |
| ![](https://github.com/JHelsen/point-centromere-detection/blob/main/PCAn_mucor/sample_outputs/plots/JBBNAA010000006.1.png) | ![](https://github.com/JHelsen/point-centromere-detection/blob/main/PCAn_mucor/sample_outputs/plots/JBBNAA010000002.1.png) |

